python/lsst/daf/butler/core/dimensions/elements.py

```diff
@@ -0,0 +1,278 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("DimensionElement", "Dimension", "DimensionJoin")
+
+from ..utils import PrivateConstructorMeta
+
+
+class DimensionElement(metaclass=PrivateConstructorMeta):
+    """Base class for elements in the dimension schema.
+
+    `DimensionElement` has exactly two subclasses: `Dimension` and
+    `DimensionJoin`.
+
+    `DimensionElement` objects are not directly constructable; they can only
+    be obtained (possibly indirectly) from a special "universe"
+    `DimensionGraph` loaded from configuration.
+    """
+
+    #
+    # Constructor is private, so its docs are just comments.
+    #
+    # Parameters
+    # ----------
+    # universe : `DimensionGraph`
+    #     Ultimate-parent `DimensionGraph` that constructed this element.
+    # name : `str`
+    #     Name of the element.  Also the name of any SQL table or view
+    #     associated with it.
+    # hasRegion : `bool`
+    #     Whether entries for this dimension are associated with a region
+    #     on the sky.
+    # link : iterable of `str`
+    #     The names of primary key fields used by this element but not any of
+    #     its dependencies.
+    # required : iterable of `str`
+    #     The names of the `Dimension`\ s whose primary keys are a subset of
+    #     this element's primary key.
```

Comment:
Yes, though I think we (perhaps unfortunately) actually have baked in that assumption to a certain degree - at least at the level of that these "link" columns must be unique identifiers and used frequently as foreign keys, and once you've gone that far I'd be surprised if there was much value to defining primary keys separately.  That said, it would be good to try to move away from the rather database-specific term `primary key` in favor of `links` or something else as much as possible.